File: python3/_2_AddTwoNumbers.py
from typing import Optional, Tuple

# Definition for singly-linked list.
from python3.helper import Eg, int_list_to_linked_list, tester_helper
from python3.helper_classes import ListNode


class Solution:
    def sum(self, l1: Optional[ListNode], l2: Optional[ListNode], carry: int) \
            -> Tuple[ListNode, int]:
        v1 = 0 if l1 is None else l1.val
        v2 = 0 if l2 is None else l2.val
        sum_ = v1 + v2 + carry
        digit = sum_ % 10
        return ListNode(digit), sum_ // 10

    def addTwoNumbers(
            self, l1: Optional[ListNode], l2: Optional[ListNode]) -> \
            Optional[ListNode]:
        head, carry = self.sum(l1, l2, 0)
        last = head
        l1 = l1.next
        l2 = l2.next
        while l1 is not None or l2 is not None or carry > 0:
            next_, carry = self.sum(l1, l2, carry)
            last.next = next_
            last = next_
            if l1 is not None:
                l1 = l1.next
            if l2 is not None:
                l2 = l2.next
        return head


def tester():
    cvt = int_list_to_linked_list
    examples = [
        Eg((cvt([2, 4, 3]), cvt([5, 6, 4])), cvt([7, 0, 8])),
        Eg((cvt([0]), cvt([0])), cvt([0])),
        Eg((cvt([9, 9, 9, 9, 9, 9, 9]), cvt([9, 9, 9, 9])),
           cvt([8, 9, 9, 9, 0, 0, 0, 1])),
    ]
    tester_helper(2, examples, Solution().addTwoNumbers)

File: rust/src/_2_add_two_numbers.rs
//! Solution for https://leetcode.com/problems/add-two-numbers
//! 2. Add Two Numbers

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(
        mut l1: Option<Box<ListNode>>,
        mut l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut has_carry = false;
        let mut result = None;
        let mut tail = &mut result;
        loop {
            (l1, l2, has_carry) = match (l1, l2, has_carry) {
                (None, None, true) => {
                    *tail = Some(Box::new(ListNode::new(1)));
                    (None, None, false)
                }
                (None, None, false) => break,
                (None, Some(x), has_carry) | (Some(x), None, has_carry) => {
                    let next_val = x.val + if has_carry { 1 } else { 0 };
                    *tail = Some(Box::new(ListNode::new(next_val % 10)));
                    tail = &mut tail.as_mut().unwrap().next;
                    (x.next, None, next_val >= 10)
                }
                (Some(x), Some(y), has_carry) => {
                    let next_val = x.val + y.val + if has_carry { 1 } else { 0 };
                    *tail = Some(Box::new(ListNode::new(next_val % 10)));
                    tail = &mut tail.as_mut().unwrap().next;
                    (x.next, y.next, next_val >= 10)
                }
            }
        }
        result
    }
}

// << ---------------- Code below here is only for local use ---------------- >>

pub struct Solution;
use cargo_leet::ListNode;

#[cfg(test)]
mod tests {
    use super::*;
    use cargo_leet::ListHead;

    use rstest::rstest;

    #[rstest]
    #[case(ListHead::from(vec![2,4,3]).into(), ListHead::from(vec![5,6,4]).into(), ListHead::from(vec![7,0,8]).into())]
    #[case(ListHead::from(vec![0]).into(), ListHead::from(vec![0]).into(), ListHead::from(vec![0]).into())]
    #[case(ListHead::from(vec![9,9,9,9,9,9,9]).into(), ListHead::from(vec![9,9,9,9]).into(), ListHead::from(vec![8,9,9,9,0,0,0,1]).into())]
    fn case(
        #[case] l1: Option<Box<ListNode>>,
        #[case] l2: Option<Box<ListNode>>,
        #[case] expected: Option<Box<ListNode>>,
    ) {
        let actual = Solution::add_two_numbers(l1, l2);
        assert_eq!(actual, expected);
    }
}
